go/vt/orchestrator/logic/topology_recovery.go

```diff
@@ -2018,3 +2028,146 @@ func GracefulMasterTakeover(clusterName string, designatedKey *inst.InstanceKey,
 
 	return topologyRecovery, promotedMasterCoordinates, err
 }
+
+// electNewMaster elects a new master while none were present before.
+func electNewMaster(analysisEntry inst.ReplicationAnalysis, candidateInstanceKey *inst.InstanceKey, forceInstanceRecovery bool, skipProcesses bool) (recoveryAttempted bool, topologyRecovery *TopologyRecovery, err error) {
+	topologyRecovery, err = AttemptRecoveryRegistration(&analysisEntry, false, true)
+	if topologyRecovery == nil {
+		AuditTopologyRecovery(topologyRecovery, fmt.Sprintf("found an active or recent recovery on %+v. Will not issue another electNewMaster.", analysisEntry.AnalyzedInstanceKey))
+		return false, nil, err
+	}
+	log.Infof("Analysis: %v, will elect a new master: %v", analysisEntry.Analysis, analysisEntry.SuggestedClusterAlias)
+
+	unlock, err := LockShard(analysisEntry.AnalyzedInstanceKey)
+	if err != nil {
+		log.Infof("CheckAndRecover: Analysis: %+v, InstanceKey: %+v, candidateInstanceKey: %+v, "+
+			"skipProcesses: %v: NOT detecting/recovering host, could not obtain shard lock (%v)",
+			analysisEntry.Analysis, analysisEntry.AnalyzedInstanceKey, candidateInstanceKey, skipProcesses, err)
+		return false, topologyRecovery, err
+	}
+	defer unlock(&err)
+
+	replicas, err := inst.ReadClusterAliasInstances(analysisEntry.SuggestedClusterAlias)
+	if err != nil {
+		return false, topologyRecovery, err
+	}
+	// TODO(sougou): this is not reliable, because of the timeout.
+	replicas = inst.StopReplicasNicely(replicas, time.Duration(config.Config.InstanceBulkOperationsWaitTimeoutSeconds)*time.Second)
+	if len(replicas) == 0 {
+		return false, topologyRecovery, fmt.Errorf("no instances in cluster %v", analysisEntry.SuggestedClusterAlias)
+	}
+
+	var candidate *inst.Instance
+	for _, replica := range replicas {
+		if candidate == nil {
+			// TODO(sougou): this needs to do more. see inst.chooseCandidateReplica
+			if !inst.IsBannedFromBeingCandidateReplica(replica) {
+				candidate = replica
+			}
+			continue
+		}
+		if err := inst.CheckMoveViaGTID(replica, candidate); err != nil {
+			if err := inst.CheckMoveViaGTID(candidate, replica); err != nil {
+				return false, topologyRecovery, fmt.Errorf("instances are not compatible: %+v %+v: %v", candidate, replica, err)
+			} else {
+				if !inst.IsBannedFromBeingCandidateReplica(replica) {
+					candidate = replica
+				}
+			}
```

Comment:
a bit too much if-else-if-else. Can you please explain the logic in words?

Reply:
I've refactored this. LMK if the new code reads better.